spikeextractors/extractors/neuroscopeextractors/neuroscopeextractors.py

```diff
@@ -0,0 +1,330 @@
+        
+        self._sampling_frequency = float(soup.samplingrate.string) # careful not to confuse it with the lfpSamplingRate
+        
+        res = np.loadtxt(resfile_path, dtype=np.int64, usecols=0, ndmin=1)
+        clu = np.loadtxt(clufile_path, dtype=np.int64, usecols=0, ndmin=1)
+        if len(res) > 0:
+            # Extract the number of clusters read as the first line of the clufile
+            # then remove it from the clu list
+            n_clu = clu[0]
+            clu = np.delete(clu, 0)
+            unique_ids = np.unique(clu)
+            
+            if not unique_ids==np.arange(n_clu+1): # some missing IDs somewhere
+                if 0 not in unique_ids: # missing unsorted IDs
+                    n_clu += 1
+                if 1 not in unique_ids: # missing mua IDs
+                    n_clu += 1
+                # If it is any other kinda of ID, then it is very strange that it is missing...
+                
+            
+            # Initialize spike trains and extract times from .res and appropriate clusters from .clu
+            # based on user input for ignoring multi-unit activity
+            self._spiketrains = []
+            if keep_mua_units: # default
+                n_clu -= 1;
+                self._unit_ids = [x+1 for x in range(n_clu)] # generates list from 1,...,clu[0]-1
+                for s_id in self._unit_ids:
+                    self._spiketrains.append(res[(clu == s_id).nonzero()])
+            else:
+                # Ignoring IDs of 0 until get_unsorted_spike_train is implemented into base
+                # Also ignoring IDs of 1 since user called keep_mua_units=False
+                n_clu -= 2;
+                self._unit_ids = [x+1 for x in range(n_clu)] # generates list from 1,...,clu[0]-2
+                for s_id in self._unit_ids:
+                    self._spiketrains.append(res[(clu == s_id+1).nonzero()]) # only reading cluster IDs 2,...,clu[0]-1
+        else:
+            self._spiketrains = []
+            self._unit_ids = []
+            
+        self._kwargs = {'resfile_path': str(Path(resfile_path).absolute()),
+                        'clufile_path': str(Path(clufile_path).absolute()),
+                        'keep_mua_units': keep_mua_units}
+
+
+    def get_unit_ids(self):
+        return list(self._unit_ids)
+    
+    
+    def get_sampling_frequency(self):
+        return self._sampling_frequency
+    
+    
+    def shift_unit_ids(self,shift):
+        self._unit_ids = [x + shift for x in self._unit_ids]
+    
+    
+    def add_unit(self, unit_id, spike_times):
+        '''This function adds a new unit with the given spike times.
+
+        Parameters
+        ----------
+        unit_id: int
+            The unit_id of the unit to be added.
+        '''
+        self._unit_ids.append(unit_id)
+        self._spiketrains.append(spike_times)
+        
+        
+    def read_shanks(general_path, shanks, keep_mua_units=True):
```

Comment:
I think we could do this systematically at the constructor level. If `res_file` and `clu_file` are used as args, the constructor can laod a single file. If instead a `folder_path` is used, we can loop through the folder, find all shanks files (suffix .res.*/.clu.*) and load them in sequence. Moreover, we can have an extra argument in the constructor, `exclude_shanks` which is `None` by default, but if an int or a list of ints are passed we can exclude those shanks from loading. Sounds good?

Reply:
Still working on this, will let you know when it's done.

Reply:
`exclude_shanks `is now incorporated into the `NeuroscopeMultiSortingExtractor`